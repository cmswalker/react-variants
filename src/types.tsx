import React from 'react';

declare global {
  interface Function {
    name: string;
  }
}

export type IVariants = {
  [key: number]: React.ComponentType
}

export interface IVariantsProps {
  variants?: IVariants;
}

export interface IWithVariantProps {
  variant?: number;
  render?: Function;
  [x: string]: any;
}

export interface IVariantCombined extends IVariantsProps, IWithVariantProps {}

export interface InternalState {
  variantCount: number;
}

export interface IVariantState extends InternalState, IVariantCombined {
  isDefault: boolean;
}

export interface IVariantProps extends IVariantState {
  displayName: string;
};

export type DefaultVariantComponent = IVariantsProps & React.ComponentType;
