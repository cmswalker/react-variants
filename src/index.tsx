import React from 'react';

import { InternalState, IVariantProps, IVariantState, DefaultVariantComponent } from './types';

const isNull = (val: any): boolean => val === null;
const isUndefined = (val: any): boolean => val === undefined;
const isFunc = (val: any): boolean => typeof val === 'function';
const isObjectLiteral = (val: any): boolean => Object.prototype.toString.call(val) === '[object Object]';
const isEmpty = (val: object): boolean => !Object.keys(val).length;
const getDisplayName = (comp : Function): string => comp.name || comp.constructor.name;
const generateRenderName = (name : string, displayName: string) : string => name || `${displayName}AsRenderProp`;
const withRenderError = (displayName: string, msg: string): Error => new Error(`Error in WithRenderProps(${displayName}): ${msg}`);
const withVariantsError = (displayName : string, msg: string) : Error => new Error(`Error in WithVariants(${displayName}): ${msg}`);

export function WithRenderProps(DefaultVariant: React.ComponentType): React.ComponentClass<any> {
  const displayName = getDisplayName(DefaultVariant);

  return class RenderWrapper extends React.Component<any, any> {
    render() {
      const topProps = this.props;
      const { render: topRender, variant = 0} = topProps;
      const isRenderProp = isFunc(topRender);
      const isDefault = !variant;
      const newProps = { ...topProps, variant, isDefault, isRenderProp, displayName };

      if (isDefault) {
        return <DefaultVariant {...newProps}  />
      }

      if (isRenderProp) {
        // @ts-ignore
        class RenderProp extends DefaultVariant<any, any> {
          render() {
            // @ts-ignore
            this.props = newProps;
            return topRender(this);
          }
        }

        // @ts-ignore
        return <RenderProp/>
      }

      throw withRenderError(displayName, 'Could not find render definition or variant');
    }
  }
}

export function WithVariants(defaultVariant: DefaultVariantComponent): React.ComponentClass<IVariantProps> {
  if (!isFunc(defaultVariant)) {
    throw new Error('Must provide initial variant as first argument');
  }

  const { variants = {}} = defaultVariant;
  variants[0] = defaultVariant;

  const displayName = getDisplayName(defaultVariant);

  if (!isObjectLiteral(variants) || isEmpty(variants)) {
    throw withVariantsError(displayName, 'Must provide variant configuration with at least one variant');
  }

  let renderVariantCount = 0;
  const staticVariantCount = Object.keys(variants).length;
  let totalVariantCount = staticVariantCount;

  return class VariantWrapper extends React.Component<IVariantProps, IVariantState> {
    constructor(props: IVariantProps) {
      super(props);

      const { variant } = props;

      if (isNull(variant) || isUndefined(variant) || isNaN(variant)) {
        throw withVariantsError(displayName, `Must provide variant prop to ${displayName}`);
      }

      const isRenderProp = isFunc(props.render);

      if (isRenderProp) {
        renderVariantCount++;
        totalVariantCount++;
      }

      const isStaticVariant = !!variants[variant];
      const initialState = {
        ...props,
        isRenderProp,
        isStaticVariant,
        displayName,
        variants: { ...variants },
        isDefault: variant === 0,
        staticVariantCount
      };

      this.state = initialState;
    }

    render() {
      const { state, props } = this;
      const { variant, variants, isRenderProp } = state;

      const VariantComponent : React.ComponentType<any> = variants[variant];
      let name: string|undefined = VariantComponent && getDisplayName(VariantComponent);

      if (isRenderProp) {
        name = generateRenderName(name, displayName);
      }

      const combinedProps = { ...state, ...props, displayName: name, renderVariantCount, totalVariantCount };

      if (isRenderProp) {
        const Rendered = WithRenderProps(defaultVariant);
        return <Rendered {...combinedProps} />;
      }

      if (!isFunc(VariantComponent)) {
        throw withVariantsError(displayName, `No variant # ${variant} exists for ${name}, check your config`);
      }

      return <VariantComponent {...combinedProps} />;
    }
  };
}
