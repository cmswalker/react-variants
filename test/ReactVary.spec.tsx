import React from 'react';
import renderer from 'react-test-renderer';

import { WithVariants, WithRenderProps } from '../src';
import { ExampleComponent, MyStatefulComponent } from '../example/exampleComponents';
import ExampleComponent1 from '../example/variants/1';
import ExampleComponent2 from '../example/variants/2';

ExampleComponent.variants = {
  1: ExampleComponent1,
  2: ExampleComponent2
}
const ExampleComponentWithVariants : React.ComponentType<any> = WithVariants(ExampleComponent);
const MyStatefulComponentWithRenderProps : React.ComponentType <any> = WithRenderProps(MyStatefulComponent);

const getTime = () => new Date().getSeconds();

beforeAll(() => {
  const constantDate = new Date('2017-06-13T04:41:20');

  // @ts-ignore
  Date = class extends Date {
    // @ts-ignore
    constructor() {
      return constantDate
    }
  }
});

afterAll(() => {

})

test('WithVariants', () => {
  const time = getTime();
  const component = renderer.create(
    <div>
      <ExampleComponentWithVariants variant={0} time={time} />
      <ExampleComponentWithVariants variant={1} time={time} />
      <ExampleComponentWithVariants variant={2} time={time} />
      <ExampleComponentWithVariants variant={3} time={time}
        render={({ time, displayName, variant, staticVariantCount, isDefault, isRenderProp, isStaticVariant, totalRenderCount, variantRenderCount }) => {
          expect(time).toBe(20);
          expect(displayName).toBe('ExampleComponentAsRenderProp');
          expect(variant).toBe(3);
          expect(staticVariantCount).toBe(3);
          expect(isDefault).toBe(false);
          expect(isRenderProp).toBe(true);
          expect(isStaticVariant).toBe(false);
          expect(variantRenderCount).toBe(1);
          return <div></div>;
      }} />
    </div>
  );

  const tree = component.toJSON();
  expect(ExampleComponent.variants[0]).toEqual(expect.any(Function));
  expect(ExampleComponent.variants[1]).toEqual(expect.any(Function));
  expect(ExampleComponent.variants[2]).toEqual(expect.any(Function));
  expect(tree).toMatchSnapshot();
});

test('WithRenderProps', () => {
  const component = renderer.create(
    <div>
      <MyStatefulComponentWithRenderProps variant={0} />
      <MyStatefulComponentWithRenderProps variant={1} render={({props, state}) => {
        // Preserve original comp state
        expect(state.date).toBe('5/2 4:20');

        expect(props.variant).toBe(1);
        expect(props.variantCount).toBe(2);
        expect(props.isDefault).toBe(false);
        expect(props.totalRenderCount).toBe(2);
        expect(props.isRenderProp).toBe(true);
        return <div></div>;
      }}/>
    </div>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});