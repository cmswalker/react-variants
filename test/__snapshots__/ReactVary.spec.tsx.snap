// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithRenderProps 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "cornflowerblue",
      }
    }
  >
    Today's Date: 
    5/2 4:20
  </div>
  <div />
</div>
`;

exports[`WithVariants 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "coral",
      }
    }
  >
    DisplayName: 
    ExampleComponent
     Variant: 
    0
     Time: 
    20
     Variant Render Count: 
     All Variants Render Count: 
  </div>
  <div
    style={
      Object {
        "backgroundColor": "turquoise",
      }
    }
  >
    DisplayName: 
    ExampleComponent1
     Variant: 
    1
     Time: 
    20
     Variant Render Count: 
  </div>
  <div
    style={
      Object {
        "backgroundColor": "cornsilk",
      }
    }
  >
    DisplayName: 
    ExampleComponent2
     Variant: 
    2
     Time: 
    20
     Variant Render Count: 
  </div>
  <div />
</div>
`;
